File: internal/c2pa/verify.go
package c2pa

import (
	"bytes"
	"crypto/ed25519"
	"crypto/sha256"
	"encoding/binary"
	"fmt"
	"hash/crc32"
)

// Reading a manifest back, and deciding whether to believe it.
//
// Four things have to hold, and each one is a way a manifest can be a lie:
//
//  1. The claim signature verifies against the key. Otherwise nothing else in
//     the manifest is evidence of anything.
//  2. Each assertion's bytes hash to what the claim says. Otherwise an
//     assertion was swapped for another under the same label.
//  3. The data hash matches the file. Otherwise the manifest describes
//     different pixels.
//  4. The excluded range is exactly where the manifest sits. Otherwise the
//     hash was computed over less than the file, and a manifest can be moved
//     onto content it never described.
//
// The fourth is the one that is easy to leave out, because a manifest without
// it still validates -- and validates against anything.

// Statement is what a verified manifest says.
type Statement struct {
	Title             string
	Format            string
	DigitalSourceType string
	SoftwareAgent     string
	Author            string
	Model             string
	Instruction       string
	When              string

	// DerivedFrom is the hash of the asset this one was made from, when the
	// manifest carries an ingredient. A caller holding the original can check
	// it; a caller who is not can at least say what this claims to be a copy
	// of.
	DerivedFrom []byte
	ParentTitle string
}

// GeneratedByModel reports whether the file was made or altered by a trained
// model, which is what the Article 50 marking obligation attaches to.
func (s Statement) GeneratedByModel() bool {
	switch s.DigitalSourceType {
	case "trainedAlgorithmicMedia", "compositeWithTrainedAlgorithmicMedia":
		return true
	}
	return false
}

// Verify reads the manifest out of a file and checks it against the file.
func Verify(file []byte, key ed25519.PublicKey) (Statement, error) {
	raw, where, err := extract(file)
	if err != nil {
		return Statement{}, err
	}

	boxes, err := parseBoxes(raw, 0)
	if err != nil {
		return Statement{}, fmt.Errorf("the manifest does not parse: %w", err)
	}
	store, err := manifestStore(boxes)
	if err != nil {
		return Statement{}, err
	}
	manifest, ok := store.find(uuidManifest)
	if !ok {
		return Statement{}, fmt.Errorf("the store holds no manifest")
	}

	claimBox, ok := manifest.find(uuidClaim)
	if !ok {
		return Statement{}, fmt.Errorf("the manifest holds no claim")
	}
	sigBox, ok := manifest.find(uuidSignature)
	if !ok {
		return Statement{}, fmt.Errorf("the manifest holds no claim signature")
	}
	assertions, ok := manifest.find(uuidAssertionStore)
	if !ok {
		return Statement{}, fmt.Errorf("the manifest holds no assertions")
	}

	claimBytes, err := claimBox.content()
	if err != nil {
		return Statement{}, err
	}
	sigBytes, err := sigBox.content()
	if err != nil {
		return Statement{}, err
	}

	// (1) The signature, first. Everything below reads values out of the
	// claim, and reading them before knowing they were signed would mean
	// acting on an attacker's numbers.
	signedClaim, err := Verify1(sigBytes, key)
	if err != nil {
		return Statement{}, err
	}
	if !bytes.Equal(signedClaim, claimBytes) {
		return Statement{}, fmt.Errorf(
			"the signature covers a different claim than the one in the " +
				"manifest, so the signed claim is not the one that would be read")
	}

	claim, err := decodeMap(claimBytes)
	if err != nil {
		return Statement{}, fmt.Errorf("the claim does not parse: %w", err)
	}

	// (2) Each assertion the claim lists, by the hash of its own bytes.
	byLabel, err := assertionBytes(assertions)
	if err != nil {
		return Statement{}, err
	}
	listed, err := checkAssertions(claim, byLabel)
	if err != nil {
		return Statement{}, err
	}

	// (3) and (4): the hard binding, and where it excludes.
	if err := checkBinding(file, byLabel[labelHash], where); err != nil {
		return Statement{}, err
	}

	return statementFrom(claim, byLabel[labelActions],
		byLabel[labelIngredient], listed)
}

// manifestStore picks the store out of the boxes at the top of the manifest.
//
// This program puts its padding inside the store, so it writes exactly one box
// here. A padding box beside the store is still tolerated, because another
// writer may reserve space that way and it carries nothing. Anything else is
// refused rather than skipped: a box a reader ignores is a place to put
// something no verifier ever looked at.
func manifestStore(boxes []parsed) (parsed, error) {
	var found parsed
	seen := false
	for _, b := range boxes {
		switch {
		case b.uuid == uuidManifestStore:
			if seen {
				return parsed{}, fmt.Errorf(
					"there are two manifest stores, so which one describes " +
						"this file depends on which is read first")
			}
			found, seen = b, true
		case b.kind == paddingKind:
			// Reserved space. Its contents are inside the excluded range.
		default:
			return parsed{}, fmt.Errorf(
				"there is a %q box beside the manifest, which this program "+
					"does not read and will not ignore", b.kind)
		}
	}
	if !seen {
		return parsed{}, fmt.Errorf("this is not a C2PA manifest store")
	}
	return found, nil
}

// where the manifest sits in the file: the range that must be excluded.
type placement struct {
	start  int
	length int
}

// extract finds the manifest store and where it sits, for whichever container
// this is.
func extract(file []byte) ([]byte, placement, error) {
	switch {
	case bytes.HasPrefix(file, pngMagic):
		return extractPNG(file)
	case bytes.HasPrefix(file, []byte{0xff, 0xd8}):
		return extractJPEG(file)
	}
	return nil, placement{}, fmt.Errorf("this is not a PNG or a JPEG")
}

func extractPNG(file []byte) ([]byte, placement, error) {
	at := len(pngMagic)
	for at+12 <= len(file) {
		length := int(binary.BigEndian.Uint32(file[at : at+4]))
		kind := string(file[at+4 : at+8])
		next := at + 12 + length
		if next <= at || next > len(file) {
			return nil, placement{}, fmt.Errorf(
				"a PNG chunk at byte %d claims %d bytes", at, length)
		}
		if kind == pngChunk {
			data := file[at+8 : at+8+length]
			// The CRC, because a chunk that fails it is a chunk a PNG reader
			// would reject, and a manifest only a lenient reader can see is
			// not a manifest the file carries.
			crc := binary.BigEndian.Uint32(file[at+8+length : next])
			if got := chunkCRC(data); got != crc {
				return nil, placement{}, fmt.Errorf(
					"the manifest chunk's CRC is %08x and it claims %08x",
					got, crc)
			}
			return data, placement{start: at, length: next - at}, nil
		}
		at = next
	}
	return nil, placement{}, fmt.Errorf("this PNG carries no manifest")
}

func extractJPEG(file []byte) ([]byte, placement, error) {
	at := 2
	for at+4 <= len(file) {
		if file[at] != 0xff {
			return nil, placement{}, fmt.Errorf("byte %d is not a marker", at)
		}
		marker := file[at+1]
		if marker < 0xe0 || marker > 0xef {
			break
		}
		length := int(binary.BigEndian.Uint16(file[at+2 : at+4]))
		next := at + 2 + length
		if length < 2 || next > len(file) {
			return nil, placement{}, fmt.Errorf(
				"a JPEG segment at byte %d claims %d bytes", at, length)
		}
		if marker == jpegAPP11 {
			payload := file[at+4 : next]
			if len(payload) < 8 {
				return nil, placement{}, fmt.Errorf(
					"an APP11 segment is too short to hold a JUMBF header")
			}
			if binary.BigEndian.Uint16(payload[:2]) != jpegCI {
				return nil, placement{}, fmt.Errorf(
					"this APP11 segment is not JUMBF")
			}
			return payload[8:], placement{start: at, length: next - at}, nil
		}
		at = next
	}
	return nil, placement{}, fmt.Errorf("this JPEG carries no manifest")
}

// chunkCRC is the PNG chunk checksum: CRC-32/IEEE over the type and the data,
// the type included because that is what the format specifies.
func chunkCRC(data []byte) uint32 {
	c := crc32.NewIEEE()
	c.Write([]byte(pngChunk))
	c.Write(data)
	return c.Sum32()
}

// checkBinding is (3) and (4): the file hashes to what the claim says, over
// exactly the bytes the manifest does not occupy.
func checkBinding(file, assertion []byte, where placement) error {
	if assertion == nil {
		return fmt.Errorf("the manifest carries no c2pa.hash.data assertion, " +
			"so it says nothing about this file's bytes and would validate " +
			"against any file at all")
	}
	m, err := decodeMap(assertion)
	if err != nil {
		return fmt.Errorf("the data hash assertion does not parse: %w", err)
	}
	if alg, _ := m["alg"].(Text); alg != "sha256" {
		return fmt.Errorf("the data hash uses %q, and this checks sha256", alg)
	}
	want, ok := m["hash"].(Bytes)
	if !ok {
		return fmt.Errorf("the data hash assertion holds no hash")
	}

	ranges, ok := m["exclusions"].(Array)
	if !ok {
		return fmt.Errorf("the data hash assertion lists no exclusions, and a " +
			"file containing its own manifest cannot hash to a value computed " +
			"over all of it")
	}

	// (4) The range has to be exactly the manifest. Not a superset: a claim
	// excluding more than it occupies is a claim over fewer bytes than the
	// file has, and those bytes can then be anything.
	if len(ranges) != 1 {
		return fmt.Errorf("the manifest claims %d excluded ranges; it occupies "+
			"one, and any other range is bytes hidden from the hash",
			len(ranges))
	}
	r, ok := ranges[0].(Map)
	if !ok {
		return fmt.Errorf("an exclusion is not a map")
	}
	start, sok := r["start"].(Uint)
	length, lok := r["length"].(Uint)
	if !sok || !lok {
		return fmt.Errorf("an exclusion has no start and length")
	}
	if int(start) != where.start || int(length) != where.length {
		return fmt.Errorf(
			"the manifest excludes bytes %d-%d and sits at %d-%d. A manifest "+
				"whose exclusion does not match where it is has a hash over "+
				"the wrong bytes, and can be moved onto content it never "+
				"described",
			start, int(start)+int(length),
			where.start, where.start+where.length)
	}

	got, err := hashExcluding(file, []exclusion{
		{start: where.start, length: where.length},
	})
	if err != nil {
		return err
	}
	if !bytes.Equal(got, want) {
		return fmt.Errorf(
			"this file hashes to %s and the manifest claims %s, so the "+
				"manifest describes different content",
			shortDigest(got), shortDigest(want))
	}
	return nil
}

// assertionBytes collects each assertion's own encoded bytes by label. The
// bytes, not the parsed value: the claim commits to a hash of these, and
// re-encoding a parsed value would compare against something else.
func assertionBytes(store parsed) (map[string][]byte, error) {
	out := map[string][]byte{}
	for _, child := range store.children {
		if child.kind != "jumb" {
			continue
		}
		body, err := child.content()
		if err != nil {
			return nil, err
		}
		if _, seen := out[child.label]; seen {
			return nil, fmt.Errorf(
				"two assertions are labelled %q, so which one the claim "+
					"commits to depends on which is read first", child.label)
		}
		out[child.label] = body
	}
	return out, nil
}

// checkAssertions is (2): every assertion the claim lists is present and
// hashes to what the claim recorded. Returns the labels checked.
func checkAssertions(claim Map, byLabel map[string][]byte) ([]string, error) {
	listed, ok := claim["assertions"].(Array)
	if !ok || len(listed) == 0 {
		return nil, fmt.Errorf("the claim lists no assertions")
	}
	var labels []string
	for _, entry := range listed {
		ref, rok := entry.(Map)
		if !rok {
			return nil, fmt.Errorf("an assertion reference is not a map")
		}
		url, uok := ref["url"].(Text)
		want, wok := ref["hash"].(Bytes)
		if !uok || !wok {
			return nil, fmt.Errorf("an assertion reference has no url and hash")
		}
		label := labelFromURI(string(url))
		body, present := byLabel[label]
		if !present {
			return nil, fmt.Errorf(
				"the claim commits to an assertion at %q and the store does "+
					"not hold it", url)
		}
		sum := sha256.Sum256(body)
		if !bytes.Equal(sum[:], want) {
			return nil, fmt.Errorf(
				"the assertion %q does not hash to what the claim signed, so "+
					"it was changed after signing", label)
		}
		labels = append(labels, label)
	}
	return labels, nil
}

// labelFromURI takes the label off a self#jumbf= reference.
func labelFromURI(uri string) string {
	if i := bytes.LastIndexByte([]byte(uri), '/'); i >= 0 {
		return uri[i+1:]
	}
	return uri
}

// statementFrom reads the human-facing values out of a claim that has already
// been verified.
func statementFrom(claim Map, actions, ingredient []byte,
	listed []string) (Statement, error) {
	s := Statement{}
	if v, ok := claim["title"].(Text); ok {
		s.Title = string(v)
	}
	if v, ok := claim["format"].(Text); ok {
		s.Format = string(v)
	}
	if v, ok := claim["claim_generator"].(Text); ok {
		s.SoftwareAgent = string(v)
	}
	if actions == nil {
		return s, nil
	}
	m, err := decodeMap(actions)
	if err != nil {
		return s, fmt.Errorf("the actions assertion does not parse: %w", err)
	}
	list, ok := m["actions"].(Array)
	if !ok || len(list) == 0 {
		return s, fmt.Errorf("the actions assertion lists no actions")
	}
	first, ok := list[0].(Map)
	if !ok {
		return s, fmt.Errorf("an action is not a map")
	}
	if v, ok := first["when"].(Text); ok {
		s.When = string(v)
	}
	if v, ok := first["digitalSourceType"].(Text); ok {
		s.DigitalSourceType = termOf(string(v))
	}
	if a, ok := first["author"].(Map); ok {
		if n, ok := a["name"].(Text); ok {
			s.Author = string(n)
		}
	}
	if p, ok := first["parameters"].(Map); ok {
		if v, ok := p["com.quilzo.model"].(Text); ok {
			s.Model = string(v)
		}
		if v, ok := p["com.quilzo.instruction"].(Text); ok {
			s.Instruction = string(v)
		}
	}

	// The ingredient, if this file says it was made from another. Read after
	// the assertion hashes were checked, so what it names was signed.
	if ingredient != nil {
		m, ierr := decodeMap(ingredient)
		if ierr != nil {
			return s, fmt.Errorf("the ingredient assertion does not parse: %w",
				ierr)
		}
		if alg, ok := m["alg"].(Text); ok && alg != "sha256" {
			return s, fmt.Errorf(
				"the ingredient names algorithm %q and this checks sha256", alg)
		}
		if h, ok := m["hash"].(Bytes); ok {
			s.DerivedFrom = append([]byte(nil), h...)
		}
		if v, ok := m["dc:title"].(Text); ok {
			s.ParentTitle = string(v)
		}
	}
	return s, nil
}

// termOf takes the bare term off an IPTC vocabulary URI, so a caller compares
// against the same strings internal/provenance stores.
func termOf(uri string) string {
	if i := bytes.LastIndexByte([]byte(uri), '/'); i >= 0 {
		return uri[i+1:]
	}
	return uri
}

func decodeMap(b []byte) (Map, error) {
	v, err := Decode(b)
	if err != nil {
		return nil, err
	}
	m, ok := v.(Map)
	if !ok {
		return nil, fmt.Errorf("this structure is not a map")
	}
	return m, nil
}

File: scripts/c2pa/verify.py
#!/usr/bin/env python3
"""An independent C2PA verifier, written against the specification.

Why this exists
---------------
The manifests are produced by Go code in internal/c2pa, and its tests are Go
tests. A reader written beside a writer repeats the writer's mistakes -- that
is not a hypothetical here, it is how the JPEG framing bug survived a passing
round-trip until exiftool disagreed.

So this parses the containers, decodes the CBOR with a third-party library,
rebuilds the COSE Sig_structure and checks the signature, recomputes the hard
binding, and follows the ingredient to the file it names. Nothing here imports
anything from the Go implementation. If the two agree, they agree for a reason.

    scripts/c2pa/verify.py <file> [--parent DIR] [--cert PEM]

Needs cbor2 and cryptography, which is why it is a script rather than part of
the program: the program has no dependencies and this is a check on it.
"""
import argparse
import hashlib
import struct
import sys

try:
    import cbor2
    from cryptography import x509
    from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
except ImportError:
    sys.exit("needs cbor2 and cryptography:  pip install cbor2 cryptography")

C2PA_SUFFIX = bytes([0x00, 0x11, 0x00, 0x10, 0x80, 0x00,
                     0x00, 0xAA, 0x00, 0x38, 0x9B, 0x71])


def uuid_for(name):
    return name.encode() + C2PA_SUFFIX


def extract(body):
    """Return (jumbf bytes, (offset, length)) for whichever container this is."""
    if body.startswith(b"\x89PNG\r\n\x1a\n"):
        at = 8
        while at + 12 <= len(body):
            length = struct.unpack(">I", body[at:at + 4])[0]
            kind = body[at + 4:at + 8]
            end = at + 12 + length
            if kind == b"caBX":
                data = body[at + 8:at + 8 + length]
                want = struct.unpack(">I", body[at + 8 + length:end])[0]
                got = 0xFFFFFFFF & __import__("zlib").crc32(kind + data)
                if got != want:
                    raise ValueError(f"chunk CRC {got:08x} claims {want:08x}")
                return data, (at, end - at)
            at = end
        raise ValueError("this PNG carries no manifest")

    if body.startswith(b"\xff\xd8"):
        at = 2
        while at + 4 <= len(body):
            if body[at] != 0xFF:
                raise ValueError(f"byte {at} is not a marker")
            marker = body[at + 1]
            if not 0xE0 <= marker <= 0xEF:
                break
            size = struct.unpack(">H", body[at + 2:at + 4])[0]
            end = at + 2 + size
            if marker == 0xEB and body[at + 4:at + 6] == b"JP":
                return body[at + 12:end], (at, end - at)
            at = end
        raise ValueError("this JPEG carries no manifest")
    raise ValueError("not a PNG or a JPEG")


def parse_boxes(data, depth=0):
    """Walk a run of JUMBF boxes into (kind, uuid, label, body, children)."""
    if depth > 16:
        raise ValueError("boxes nest too deeply")
    out, at = [], 0
    while at < len(data):
        if len(data) - at < 8:
            raise ValueError("trailing bytes shorter than a box header")
        length = struct.unpack(">I", data[at:at + 4])[0]
        kind = data[at + 4:at + 8].decode("latin1")
        if length < 8 or at + length > len(data):
            raise ValueError(f"a box claims {length} bytes")
        payload = data[at + 8:at + length]
        uuid = label = None
        children = []
        body = None
        if kind == "jumb":
            children = parse_boxes(payload, depth + 1)
            if not children or children[0]["kind"] != "jumd":
                raise ValueError("a superbox does not open with a description")
            d = children[0]["body"]
            uuid = d[:16]
            if d[16] & 0x02:
                label = d[17:d.index(b"\0", 17)].decode()
            children = children[1:]
        else:
            body = payload
        out.append({"kind": kind, "uuid": uuid, "label": label,
                    "body": body, "children": children})
        at += length
    return out


def find(box, uuid):
    for c in box["children"]:
        if c["uuid"] == uuid:
            return c
    return None


def content(box):
    for c in box["children"]:
        if c["kind"] in ("cbor", "json"):
            return c["body"]
    raise ValueError(f"box {box['label']!r} carries no content")


def hash_excluding(body, start, length):
    h = hashlib.sha256()
    h.update(body[:start])
    h.update(body[start + length:])
    return h.digest()


def check(path, parent_dir=None, cert_pem=None):
    body = open(path, "rb").read()
    raw, (offset, length) = extract(body)

    boxes = parse_boxes(raw)
    store = next((b for b in boxes if b["uuid"] == uuid_for("c2pa")), None)
    if store is None:
        raise ValueError("no manifest store")
    for b in boxes:
        if b["uuid"] != uuid_for("c2pa") and b["kind"] != "free":
            raise ValueError(f"unexpected {b['kind']!r} box beside the manifest")

    manifest = find(store, uuid_for("c2ma"))
    claim_box = find(manifest, uuid_for("c2cl"))
    sig_box = find(manifest, uuid_for("c2cs"))
    assertions = find(manifest, uuid_for("c2as"))

    claim_bytes = content(claim_box)
    sig_bytes = content(sig_box)
    claim = cbor2.loads(claim_bytes)

    # -- the signature -------------------------------------------------------
    cose = cbor2.loads(sig_bytes)
    tag = None
    if isinstance(cose, cbor2.CBORTag):
        tag, cose = cose.tag, cose.value
    if tag != 18:
        raise ValueError(f"COSE tag is {tag}, expected 18")
    protected, unprotected, payload, signature = cose
    headers = cbor2.loads(protected)
    alg = headers.get(1)
    if alg != -8:
        raise ValueError(f"algorithm {alg} is not EdDSA")
    if payload != claim_bytes:
        raise ValueError("the signature covers a different claim")

    der = headers.get(33)
    if isinstance(der, list):
        der = der[0]
    if cert_pem:
        cert = x509.load_pem_x509_certificate(open(cert_pem, "rb").read())
    else:
        cert = x509.load_der_x509_certificate(der)
    pub = cert.public_key()
    if not isinstance(pub, Ed25519PublicKey):
        raise ValueError("the certificate does not hold an Ed25519 key")

    sig_structure = cbor2.dumps(["Signature1", protected, b"", payload],
                                canonical=True)
    pub.verify(signature, sig_structure)

    # -- the assertions ------------------------------------------------------
    by_label = {}
    for c in assertions["children"]:
        if c["kind"] == "jumb":
            by_label[c["label"]] = content(c)

    for ref in claim["assertions"]:
        label = ref["url"].rsplit("/", 1)[-1]
        if label not in by_label:
            raise ValueError(f"the claim commits to {label} and it is absent")
        got = hashlib.sha256(by_label[label]).digest()
        if got != ref["hash"]:
            raise ValueError(f"assertion {label} does not match what was signed")

    # -- the hard binding ----------------------------------------------------
    data_hash = cbor2.loads(by_label["c2pa.hash.data"])
    ex = data_hash["exclusions"]
    if len(ex) != 1:
        raise ValueError(f"{len(ex)} exclusions; the manifest occupies one")
    if ex[0]["start"] != offset or ex[0]["length"] != length:
        raise ValueError(
            f"the manifest excludes {ex[0]['start']}..{ex[0]['start'] + ex[0]['length']} "
            f"and sits at {offset}..{offset + length}")
    got = hash_excluding(body, offset, length)
    if got != data_hash["hash"]:
        raise ValueError("the file does not hash to what the manifest claims")

    # -- what it says --------------------------------------------------------
    actions = cbor2.loads(by_label["c2pa.actions.v2"])["actions"][0]
    result = {
        "title": claim.get("title"),
        "generator": claim.get("claim_generator"),
        "action": actions.get("action"),
        "source_type": (actions.get("digitalSourceType") or "").rsplit("/", 1)[-1],
        "model": (actions.get("parameters") or {}).get("com.quilzo.model"),
        "signature": "verified (Ed25519 over COSE Sig_structure)",
        "binding": f"verified over {len(body) - length} bytes",
    }

    # -- the claim binding ---------------------------------------------------
    if "c2pa.ingredient.v3" in by_label:
        ing = cbor2.loads(by_label["c2pa.ingredient.v3"])
        result["derived_from"] = ing["hash"].hex()
        result["parent_title"] = ing.get("dc:title")
        result["relationship"] = ing.get("relationship")
        if parent_dir:
            import os
            found = None
            for name in os.listdir(parent_dir):
                p = os.path.join(parent_dir, name)
                if not os.path.isfile(p):
                    continue
                if hashlib.sha256(open(p, "rb").read()).digest() == ing["hash"]:
                    found = name
                    break
            result["parent_resolved"] = found or "NOT FOUND in " + parent_dir
    return result


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("file")
    ap.add_argument("--parent", help="directory to resolve the ingredient in")
    ap.add_argument("--cert", help="verify against this PEM rather than the embedded chain")
    a = ap.parse_args()
    try:
        for k, v in check(a.file, a.parent, a.cert).items():
            print(f"  {k:16} {v}")
        print("\nVERIFIED")
    except Exception as e:
        print(f"REFUSED: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()
